src/server/networkTools.cpp

```diff
@@ -0,0 +1,137 @@
+    std::string interfaceIp = std::string(inet_ntoa(pAddress->sin_addr));
+    interfaces.insert(std::pair<std::string, std::string>(interfaceName, interfaceIp));
+  }
+#else
+  /* Get Network interfaces information */
+  char buf[16384];
+  struct ifconf ifconf;
+  int fd = socket(PF_INET, SOCK_DGRAM, 0); /* Only IPV4 */
+  ifconf.ifc_len=sizeof buf;
+  ifconf.ifc_buf=buf;
+  if(ioctl(fd, SIOCGIFCONF, &ifconf)!=0) {
+    perror("ioctl(SIOCGIFCONF)");
+    exit(EXIT_FAILURE);
+  }
+
+  /* Go through each interface */
+  int i;
+  size_t len;
+  struct ifreq *ifreq;
+  ifreq = ifconf.ifc_req;
+  for (i = 0; i < ifconf.ifc_len; ) {
+    if (ifreq->ifr_addr.sa_family == AF_INET) {
+      /* Get the network interface ip */
+      char host[128] = { 0 };
+      const int error = getnameinfo(&(ifreq->ifr_addr), sizeof ifreq->ifr_addr,
+                                    host, sizeof host,
+                                    0, 0, NI_NUMERICHOST);
+      if (!error) {
+        std::string interfaceName = std::string(ifreq->ifr_name);
+        std::string interfaceIp = std::string(host);
+        /* Add to the map */
+        interfaces.insert(std::pair<std::string, std::string>(interfaceName, interfaceIp));
+      } else {
+        perror("getnameinfo()");
+      }
+    }
+
+    /* some systems have ifr_addr.sa_len and adjust the length that
+     * way, but not mine. weird */
+#ifndef __linux__
+    len=IFNAMSIZ + ifreq->ifr_addr.sa_len;
+#else
+    len=sizeof *ifreq;
+#endif
+    ifreq=(struct ifreq*)((char*)ifreq+len);
+    i+=len;
+  }
+#endif
+  return interfaces;
+}
+
+std::string kiwix::getBestPublicIp() {
+  std::map<std::string, std::string> interfaces = kiwix::getNetworkInterfaces();
+
+#ifndef _WIN32
+  const char* const prioritizedNames[] =
+      { "eth0", "eth1", "wlan0", "wlan1", "en0", "en1" };
+  const int count = (sizeof prioritizedNames) / (sizeof prioritizedNames[0]);
+  for (int i = 0; i < count; ++i) {
+    std::map<std::string, std::string>::const_iterator it =
+        interfaces.find(prioritizedNames[i]);
+    if (it != interfaces.end())
+      return it->second;
+  }
+#endif
+
+  for (std::map<std::string, std::string>::iterator iter = interfaces.begin();
+      iter != interfaces.end(); ++iter) {
+    std::string interfaceIp = iter->second;
+    if (interfaceIp.length() >= 7 && interfaceIp.substr(0, 7) == "192.168")
+      return interfaceIp;
+  }
+
+  for (std::map<std::string, std::string>::iterator iter = interfaces.begin();
+      iter != interfaces.end(); ++iter) {
+    std::string interfaceIp = iter->second;
+    if (interfaceIp.length() >= 7 && interfaceIp.substr(0, 7) == "172.16.")
+      return interfaceIp;
+  }
+
+  for (std::map<std::string, std::string>::iterator iter = interfaces.begin();
+      iter != interfaces.end(); ++iter) {
+    std::string interfaceIp = iter->second;
+    if (interfaceIp.length() >= 3 && interfaceIp.substr(0, 3) == "10.")
+      return interfaceIp;
+  }
```

Comment:
The three independent loop could be combine in two imbricated loop (one looping on 3 prefix and the child loop going on the interfaces)